File: apps/api/src/rag/chunker.ts
/**
 * Markdown chunking for the knowledge base.
 * Chunk size 500-1000 estimated tokens with 100-token overlap (chars/4 heuristic).
 * Splits follow markdown sections (##) and avoids cutting mid-sentence.
 */

export interface RAGChunk {
  text: string;
  section: string;
}

export interface ChunkerOptions {
  /** Upper bound of estimated tokens per chunk. */
  maxTokens?: number;
  /** Estimated tokens carried over between consecutive chunks. */
  overlapTokens?: number;
  /** Estimated tokens a chunk should reach before being flushed. */
  minTokens?: number;
}

const DEFAULT_OPTIONS: Required<ChunkerOptions> = {
  maxTokens: 1000,
  overlapTokens: 100,
  minTokens: 500,
};

/** Approximate tokens via ~4 chars per token (enough for chunk bounds). */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

/** Split markdown into level-2 (##) sections. */
export function splitSections(markdown: string): { header: string; body: string }[] {
  const sections: { header: string; body: string[] }[] = [];
  for (const line of markdown.split('\n')) {
    const m = /^#{2,}\s+(.*)$/.exec(line);
    if (m) {
      sections.push({ header: m[1].trim(), body: [] });
    } else if (sections.length > 0) {
      const last = sections[sections.length - 1];
      if (last) last.body.push(line);
    }
  }
  return sections.map((s) => ({ header: s.header, body: s.body.join('\n').trim() }));
}

/** Split text into sentence-ish units, never cutting mid-sentence. */
export function splitSentences(text: string): string[] {
  return text
    .replace(/\r\n/g, '\n')
    .split(/\n+/) // keep bullet/list lines intact
    .flatMap((line) => line.split(/(?<=[.!?])\s+/))
    .map((s) => s.trim())
    .filter(Boolean);
}

/**
 * Chunk a whole markdown document.
 * Each chunk stays within maxTokens and carries ~overlapTokens from the
 * previous chunk. Sections are kept separate (overlap resets between them).
 */
export function chunkMarkdown(markdown: string, options: ChunkerOptions = {}): RAGChunk[] {
  const { maxTokens, overlapTokens, minTokens } = { ...DEFAULT_OPTIONS, ...options };
  const chunks: RAGChunk[] = [];

  for (const section of splitSections(markdown)) {
    if (!section.header.trim()) continue;
    const sentences = splitSentences(section.body);
    if (sentences.length === 0) continue;

    let current: string[] = [];
    let overlapParts: string[] = [];

    const flush = () => {
      const text = [...overlapParts, ...current].join(' ').trim();
      if (!text) return;
      const full = text.length;
      chunks.push({ text, section: section.header });
      overlapParts = [];
      let carry = '';
      for (let i = full - 1; i >= 0 && estimateTokens(carry) < overlapTokens; i--) {
        carry = `${text[i] ?? ''}${carry}`;
      }
      overlapParts = carry !== '' ? carry.trim().split(' ').filter(Boolean) : [];
      current = [];
    };

    for (const sentence of sentences) {
      const candidate = [...overlapParts, ...current, sentence].join(' ');
      if (
        current.length > 0 &&
        estimateTokens(candidate) > maxTokens &&
        estimateTokens([...overlapParts, ...current].join(' ')) >= minTokens
      ) {
        flush();
      }
      current.push(sentence);
      if (estimateTokens([...overlapParts, ...current].join(' ')) >= maxTokens) {
        flush();
      }
    }
    flush();
  }

  // Fallback for markdown without any ## section.
  if (chunks.length === 0) {
    const sentences = splitSentences(markdown);
    let buff: string[] = [];
    for (const sentence of sentences) {
      buff.push(sentence);
      if (estimateTokens(buff.join(' ')) >= minTokens) {
        chunks.push({ text: buff.join(' '), section: 'Dokumen' });
        buff = [];
      }
    }
    if (buff.length) chunks.push({ text: buff.join(' '), section: 'Dokumen' });
  }

  return chunks;
}

File: apps/ai/app/rag/chunker.py
"""Markdown chunking for the knowledge base.

Port of `apps/api/src/rag/chunker.ts` — chunk size 500-1000 estimated tokens
with 100-token overlap (chars/4 heuristic). Splits follow markdown sections
(##) and avoids cutting mid-sentence.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional

DEFAULT_OPTIONS = {"maxTokens": 1000, "overlapTokens": 100, "minTokens": 500}


@dataclass
class RAGChunk:
    text: str
    section: str


def estimate_tokens(text: str) -> int:
    """Approximate tokens via ~4 chars per token (enough for chunk bounds)."""
    return (len(text) + 3) // 4


def split_sections(markdown: str) -> list[dict]:
    """Split markdown into level-2 (##) sections."""
    sections: list[dict] = []
    for line in markdown.split("\n"):
        m = re.match(r"^#{2,}\s+(.*)$", line)
        if m:
            sections.append({"header": m.group(1).strip(), "body": []})
        elif sections:
            sections[-1]["body"].append(line)
    return [{"header": s["header"], "body": "\n".join(s["body"]).strip()} for s in sections]


def split_sentences(text: str) -> list[str]:
    """Split text into sentence-ish units, never cutting mid-sentence."""
    parts: list[str] = []
    for line in text.replace("\r\n", "\n").split("\n"):
        for seg in re.split(r"(?<=[.!?])\s+", line):
            seg = seg.strip()
            if seg:
                parts.append(seg)
    return parts


def chunk_markdown(markdown: str, options: Optional[dict] = None) -> list[RAGChunk]:
    """Chunk a whole markdown document. Each chunk stays within maxTokens and
    carries ~overlapTokens from the previous chunk. Sections stay separate."""
    opts = {**DEFAULT_OPTIONS, **(options or {})}
    max_tokens = opts["maxTokens"]
    overlap_tokens = opts["overlapTokens"]
    min_tokens = opts["minTokens"]
    chunks: list[RAGChunk] = []

    def flush_body(
        header: str,
        current: list[str],
        overlap_parts: list[str],
    ) -> tuple[list[RAGChunk], list[str]]:
        text = " ".join([*overlap_parts, *current]).strip()
        added: list[RAGChunk] = []
        if text:
            full = len(text)
            added.append(RAGChunk(text=text, section=header))
            # carry overlap (measured in estimated tokens)
            carry = ""
            for i in range(full - 1, -1, -1):
                if estimate_tokens(carry) >= overlap_tokens:
                    break
                carry = text[i] + carry
            next_overlap = carry.strip().split(" ") if carry.strip() else []
            return added, next_overlap
        return added, []

    for section in split_sections(markdown):
        header = section["header"]
        if not header.strip():
            continue
        sentences = split_sentences(section["body"])
        if not sentences:
            continue

        current: list[str] = []
        overlap_parts: list[str] = []

        def cur_text() -> str:
            return " ".join([*overlap_parts, *current])

        for sentence in sentences:
            candidate = " ".join([*overlap_parts, *current, sentence])
            if (
                current
                and estimate_tokens(candidate) > max_tokens
                and estimate_tokens(cur_text()) >= min_tokens
            ):
                emitted, overlap_parts = flush_body(header, current, overlap_parts)
                chunks.extend(emitted)
                current = []
            current.append(sentence)
            if estimate_tokens(cur_text()) >= max_tokens:
                emitted, overlap_parts = flush_body(header, current, overlap_parts)
                chunks.extend(emitted)
                current = []
        emitted, _ = flush_body(header, current, overlap_parts)
        chunks.extend(emitted)

    # Fallback for markdown without any ## section.
    if not chunks:
        sentences = split_sentences(markdown)
        buff: list[str] = []
        for sentence in sentences:
            buff.append(sentence)
            if estimate_tokens(" ".join(buff)) >= min_tokens:
                chunks.append(RAGChunk(text=" ".join(buff), section="Dokumen"))
                buff = []
        if buff:
            chunks.append(RAGChunk(text=" ".join(buff), section="Dokumen"))

    return chunks